File: FlexServ Backend/SpringBoot/TransactionService/src/main/java/com/flexserv/service/TransactionServiceImpl.java
package com.flexserv.service;

import java.math.BigDecimal;
import java.util.List;
import java.util.stream.Collectors;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.flexserv.dto.request.BookingCreateRequest;
import com.flexserv.dto.request.BookingStatusUpdateRequest;
import com.flexserv.dto.response.BookingResponse;
import com.flexserv.entity.Address;
import com.flexserv.entity.Booking;
import com.flexserv.entity.BookingStatus;
import com.flexserv.entity.ServiceProvider;
import com.flexserv.entity.User;
import com.flexserv.exception.InvalidBookingStatusException;
import com.flexserv.exception.ResourceNotFoundException;
import com.flexserv.repository.AddressRepository;
import com.flexserv.repository.BookingRepository;
import com.flexserv.repository.ServiceProviderRepository;
import com.flexserv.repository.ServiceRepository;
import com.flexserv.repository.UserRepository;

import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
public class TransactionServiceImpl implements TransactionService {

    private final BookingRepository bookingRepository;
    private final UserRepository userRepository;
    private final ServiceProviderRepository serviceProviderRepository;
    private final ServiceRepository serviceRepository;
    private final AddressRepository addressRepository;

    @Override
    @Transactional
    public BookingResponse createBooking(BookingCreateRequest request) {
        User customer = userRepository.findById(request.getCustomerId())
                .orElseThrow(() -> new ResourceNotFoundException("Customer not found with ID: " + request.getCustomerId()));

        ServiceProvider provider = serviceProviderRepository.findById(request.getProviderId())
                .orElseThrow(() -> new ResourceNotFoundException("Service Provider not found with ID: " + request.getProviderId()));

        com.flexserv.entity.Service service = serviceRepository.findById(request.getServiceId())
                .orElseThrow(() -> new ResourceNotFoundException("Service not found with ID: " + request.getServiceId()));

        Address address = addressRepository.findById(request.getAddressId())
                .orElseThrow(() -> new ResourceNotFoundException("Address not found with ID: " + request.getAddressId()));

        BigDecimal price = (request.getTotalPrice() != null) ? request.getTotalPrice() : service.getPrice();

        Booking booking = new Booking();
        booking.setCustomer(customer);
        booking.setProvider(provider);
        booking.setService(service);
        booking.setAddress(address);
        booking.setBookingDate(request.getBookingDate());
        booking.setBookingTime(request.getBookingTime());
        booking.setTotalPrice(price);
        booking.setStatus(BookingStatus.PENDING.name()); // Lifecycle start: PENDING

        Booking savedBooking = bookingRepository.save(booking);
        return mapToResponse(savedBooking);
    }

    @Override
    public BookingResponse getBookingById(Long id) {
        Booking booking = bookingRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Booking not found with ID: " + id));
        return mapToResponse(booking);
    }

    @Override
    public List<BookingResponse> getAllBookings() {
        return bookingRepository.findAll().stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
    }

    @Override
    public List<BookingResponse> getBookingsByCustomer(Long customerId) {
        return bookingRepository.findByCustomerId(customerId).stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
    }

    @Override
    public List<BookingResponse> getBookingsByProvider(Long providerId) {
        return bookingRepository.findByProviderId(providerId).stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
    }

    @Override
    public List<BookingResponse> getBookingsByStatus(String status) {
        return bookingRepository.findByStatus(status.toUpperCase()).stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
    }

    @Override
    @Transactional
    public BookingResponse updateBookingStatus(Long bookingId, BookingStatusUpdateRequest request) {
        Booking booking = bookingRepository.findById(bookingId)
                .orElseThrow(() -> new ResourceNotFoundException("Booking not found with ID: " + bookingId));

        String targetStatusStr = request.getStatus().toUpperCase();

        if (!BookingStatus.isValidStatus(targetStatusStr)) {
            throw new InvalidBookingStatusException("Invalid status: " + request.getStatus() + ". Allowed: PENDING, CONFIRMED, IN_PROGRESS, COMPLETED, CANCELLED, REJECTED");
        }

        validateStateTransition(booking.getStatus(), targetStatusStr);

        booking.setStatus(targetStatusStr);
        Booking updatedBooking = bookingRepository.save(booking);
        return mapToResponse(updatedBooking);
    }

    @Override
    @Transactional
    public BookingResponse confirmBooking(Long bookingId) {
        return transitionStatus(bookingId, BookingStatus.CONFIRMED.name());
    }

    @Override
    @Transactional
    public BookingResponse startBookingService(Long bookingId) {
        return transitionStatus(bookingId, BookingStatus.IN_PROGRESS.name());
    }

    @Override
    @Transactional
    public BookingResponse completeBooking(Long bookingId) {
        return transitionStatus(bookingId, BookingStatus.COMPLETED.name());
    }

    @Override
    @Transactional
    public BookingResponse cancelBooking(Long bookingId, String reason) {
        return transitionStatus(bookingId, BookingStatus.CANCELLED.name());
    }

    @Override
    @Transactional
    public BookingResponse rejectBooking(Long bookingId, String reason) {
        return transitionStatus(bookingId, BookingStatus.REJECTED.name());
    }

    private BookingResponse transitionStatus(Long bookingId, String targetStatusStr) {
        Booking booking = bookingRepository.findById(bookingId)
                .orElseThrow(() -> new ResourceNotFoundException("Booking not found with ID: " + bookingId));

        validateStateTransition(booking.getStatus(), targetStatusStr);
        booking.setStatus(targetStatusStr);
        return mapToResponse(bookingRepository.save(booking));
    }

    private void validateStateTransition(String currentStatus, String newStatus) {
        if (currentStatus == null) return;
        String current = currentStatus.toUpperCase();

        if (current.equals(newStatus)) return; // No change

        if (current.equals("COMPLETED") || current.equals("CANCELLED") || current.equals("REJECTED")) {
            throw new InvalidBookingStatusException("Cannot transition terminal booking status '" + current + "' to '" + newStatus + "'");
        }
    }

    private BookingResponse mapToResponse(Booking booking) {
        String addressDetails = (booking.getAddress() != null)
                ? String.format("%s, %s, %s %s",
                booking.getAddress().getStreetAddress(),
                booking.getAddress().getCity(),
                booking.getAddress().getState(),
                booking.getAddress().getZipCode())
                : "";

        return new BookingResponse(
                booking.getId(),
                booking.getCustomer() != null ? booking.getCustomer().getId() : null,
                booking.getCustomer() != null ? booking.getCustomer().getName() : "",
                booking.getCustomer() != null ? booking.getCustomer().getEmail() : "",
                booking.getProvider() != null ? booking.getProvider().getId() : null,
                booking.getProvider() != null ? booking.getProvider().getCompanyName() : "",
                booking.getService() != null ? booking.getService().getId() : null,
                booking.getService() != null ? booking.getService().getName() : "",
                booking.getAddress() != null ? booking.getAddress().getId() : null,
                addressDetails,
                booking.getBookingDate(),
                booking.getBookingTime(),
                booking.getStatus(),
                booking.getTotalPrice(),
                booking.getCreatedAt()
        );
    }
}

File: FlexServ Backend/DotNet/TransactionService/Services/TransactionServiceImpl.cs
using Microsoft.EntityFrameworkCore;
using TransactionService.Data;
using TransactionService.Dtos;
using TransactionService.Exceptions;
using TransactionService.Models;

namespace TransactionService.Services;

public class TransactionServiceImpl : ITransactionService
{
    private readonly AppDbContext _context;

    public TransactionServiceImpl(AppDbContext context)
    {
        _context = context;
    }

    public async Task<BookingResponse> CreateBookingAsync(BookingCreateRequest request)
    {
        var customer = await _context.Users.FindAsync(request.CustomerId)
            ?? throw new ResourceNotFoundException("Customer not found with ID: " + request.CustomerId);

        var provider = await _context.ServiceProviders.FindAsync(request.ProviderId)
            ?? throw new ResourceNotFoundException("Service Provider not found with ID: " + request.ProviderId);

        var service = await _context.Services.FindAsync(request.ServiceId)
            ?? throw new ResourceNotFoundException("Service not found with ID: " + request.ServiceId);

        var address = await _context.Addresses.FindAsync(request.AddressId)
            ?? throw new ResourceNotFoundException("Address not found with ID: " + request.AddressId);

        decimal price = request.TotalPrice ?? service.Price;

        var booking = new Booking
        {
            CustomerId = customer.Id,
            Customer = customer,
            ProviderId = provider.Id,
            Provider = provider,
            ServiceId = service.Id,
            Service = service,
            AddressId = address.Id,
            Address = address,
            BookingDate = request.BookingDate,
            BookingTime = request.BookingTime,
            TotalPrice = price,
            Status = BookingStatus.PENDING.ToString(),
            CreatedAt = DateTime.UtcNow
        };

        _context.Bookings.Add(booking);
        await _context.SaveChangesAsync();

        return MapToResponse(booking);
    }

    public async Task<BookingResponse> GetBookingByIdAsync(long id)
    {
        var booking = await GetBookingWithDetailsQuery()
            .FirstOrDefaultAsync(b => b.Id == id)
            ?? throw new ResourceNotFoundException("Booking not found with ID: " + id);

        return MapToResponse(booking);
    }

    public async Task<List<BookingResponse>> GetAllBookingsAsync()
    {
        var bookings = await GetBookingWithDetailsQuery().ToListAsync();
        return bookings.Select(MapToResponse).ToList();
    }

    public async Task<List<BookingResponse>> GetBookingsByCustomerAsync(long customerId)
    {
        var bookings = await GetBookingWithDetailsQuery()
            .Where(b => b.CustomerId == customerId)
            .ToListAsync();

        return bookings.Select(MapToResponse).ToList();
    }

    public async Task<List<BookingResponse>> GetBookingsByProviderAsync(long providerId)
    {
        var bookings = await GetBookingWithDetailsQuery()
            .Where(b => b.ProviderId == providerId)
            .ToListAsync();

        return bookings.Select(MapToResponse).ToList();
    }

    public async Task<List<BookingResponse>> GetBookingsByStatusAsync(string status)
    {
        string upperStatus = status.ToUpper();
        var bookings = await GetBookingWithDetailsQuery()
            .Where(b => b.Status.ToUpper() == upperStatus)
            .ToListAsync();

        return bookings.Select(MapToResponse).ToList();
    }

    public async Task<BookingResponse> UpdateBookingStatusAsync(long bookingId, BookingStatusUpdateRequest request)
    {
        var booking = await GetBookingWithDetailsQuery()
            .FirstOrDefaultAsync(b => b.Id == bookingId)
            ?? throw new ResourceNotFoundException("Booking not found with ID: " + bookingId);

        string targetStatusStr = request.Status.ToUpper();

        if (!BookingStatusExtensions.IsValidStatus(targetStatusStr))
        {
            throw new InvalidBookingStatusException(
                "Invalid status: " + request.Status + ". Allowed: PENDING, CONFIRMED, IN_PROGRESS, COMPLETED, CANCELLED, REJECTED");
        }

        ValidateStateTransition(booking.Status, targetStatusStr);

        booking.Status = targetStatusStr;
        _context.Bookings.Update(booking);
        await _context.SaveChangesAsync();

        return MapToResponse(booking);
    }

    public async Task<BookingResponse> ConfirmBookingAsync(long bookingId)
    {
        return await TransitionStatusAsync(bookingId, BookingStatus.CONFIRMED.ToString());
    }

    public async Task<BookingResponse> StartBookingServiceAsync(long bookingId)
    {
        return await TransitionStatusAsync(bookingId, BookingStatus.IN_PROGRESS.ToString());
    }

    public async Task<BookingResponse> CompleteBookingAsync(long bookingId)
    {
        return await TransitionStatusAsync(bookingId, BookingStatus.COMPLETED.ToString());
    }

    public async Task<BookingResponse> CancelBookingAsync(long bookingId, string? reason)
    {
        return await TransitionStatusAsync(bookingId, BookingStatus.CANCELLED.ToString());
    }

    public async Task<BookingResponse> RejectBookingAsync(long bookingId, string? reason)
    {
        return await TransitionStatusAsync(bookingId, BookingStatus.REJECTED.ToString());
    }

    private async Task<BookingResponse> TransitionStatusAsync(long bookingId, string targetStatusStr)
    {
        var booking = await GetBookingWithDetailsQuery()
            .FirstOrDefaultAsync(b => b.Id == bookingId)
            ?? throw new ResourceNotFoundException("Booking not found with ID: " + bookingId);

        ValidateStateTransition(booking.Status, targetStatusStr);

        booking.Status = targetStatusStr;
        _context.Bookings.Update(booking);
        await _context.SaveChangesAsync();

        return MapToResponse(booking);
    }

    private void ValidateStateTransition(string? currentStatus, string newStatus)
    {
        if (currentStatus == null) return;
        string current = currentStatus.ToUpper();

        if (string.Equals(current, newStatus, StringComparison.OrdinalIgnoreCase)) return; // No change

        if (current == "COMPLETED" || current == "CANCELLED" || current == "REJECTED")
        {
            throw new InvalidBookingStatusException($"Cannot transition terminal booking status '{current}' to '{newStatus}'");
        }
    }

    private IQueryable<Booking> GetBookingWithDetailsQuery()
    {
        return _context.Bookings
            .Include(b => b.Customer)
            .Include(b => b.Provider)
            .Include(b => b.Service)
            .Include(b => b.Address);
    }

    private BookingResponse MapToResponse(Booking booking)
    {
        string addressDetails = "";
        if (booking.Address != null)
        {
            var parts = new List<string?> { booking.Address.HouseNo, booking.Address.Street, booking.Address.Area, booking.Address.City, booking.Address.State, booking.Address.Pincode }
                .Where(s => !string.IsNullOrWhiteSpace(s));
            addressDetails = string.Join(", ", parts);
        }

        return new BookingResponse(
            booking.Id,
            booking.Customer?.Id,
            booking.Customer?.Name ?? "",
            booking.Customer?.Email ?? "",
            booking.Provider?.Id,
            booking.Provider?.CompanyName ?? "",
            booking.Service?.Id,
            booking.Service?.Name ?? "",
            booking.Address?.Id,
            addressDetails,
            booking.BookingDate,
            booking.BookingTime,
            booking.Status,
            booking.TotalPrice,
            booking.CreatedAt
        );
    }
}
